FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift

```diff
@@ -38,19 +38,73 @@ extension FetchedMessage {
 }
 
 // MARK: - MessageService
+enum MessageServiceError: Error {
+    case missedPassPhrase(_ rawMimeData: Data)
+    case wrongPassPhrase(_ rawMimeData: Data, _ passPhrase: String)
+    // Could not fetch keys
+    case emptyKeys
+}
+
 final class MessageService {
     private let messageProvider: MessageProvider
-    private let dataService: DataServiceType & KeyDataServiceType
+    private let keyService: KeyServiceType
+    private let passPhraseStorage: PassPhraseStorageType
     private let core: Core
 
     init(
         messageProvider: MessageProvider = MailProvider.shared.messageProvider,
-        dataService: DataServiceType & KeyDataServiceType = DataService.shared,
-        core: Core = Core.shared
+        keyService: KeyServiceType = KeyService(),
+        core: Core = Core.shared,
+        passPhraseStorage: PassPhraseStorageType = PassPhraseStorage(
+            storage: EncryptedStorage(),
+            emailProvider: DataService.shared
+        )
     ) {
         self.messageProvider = messageProvider
-        self.dataService = dataService
+        self.keyService = keyService
         self.core = core
+        self.passPhraseStorage = passPhraseStorage
+    }
+
+    func validateMessage(rawMimeData: Data, with passPhrase: String) -> Promise<FetchedMessage> {
+        Promise<FetchedMessage> { [weak self] resolve, reject in
+            guard let self = self else { return }
+
+            guard let keys = try? self.keyService.getPrivateKeys(with: passPhrase).get(), keys.isNotEmpty else {
+                return reject(MessageServiceError.emptyKeys)
+            }
+
+            // TODO: - Tom - is it possible to get longid of the key which was used for decryption?
+            let decrypted = try self.core.parseDecryptMsg(
+                encrypted: rawMimeData,
+                keys: keys,
+                msgPwd: nil,
+                isEmail: true
+            )
+
+            let isDecryptError = decrypted.blocks.isAnyError
+
+            if isDecryptError {
+                reject(MessageServiceError.wrongPassPhrase(rawMimeData, passPhrase))
```

Comment:
This is naive error handling - we need to check for a particular error type, decrypt errors can happen due to various different things. See here:

```
        enum ErrorType: String, Decodable {
            case keyMismatch = "key_mismatch"
            case usePassword = "use_password"
            case wrongPwd = "wrong_password"
            case noMdc = "no_mdc"
            case badMdc = "bad_mdc"
            case needPassphrase = "need_passphrase"
            case format
            case other
        }
```

If you are rejecting for `reject(MessageServiceError.wrongPassPhrase(rawMimeData, passPhrase))` then you need to check specifically for that specific error. I think in this case it would be `needPassphrase` but I'm not 100% sure, please test it (seems like I should introduce a `wrongPassPhrase` error type, I thought we already had one).

Reply:
If this is a wrong pass phrase then you can render a pass phrase input modal, but if it's other type of error than you should render that error in detail as it was done before.

Reply:
But why? 
```
func savePassPhrase(with passPhrase: PassPhrase, inStorage: Bool) {
        if inStorage {
            logger.logInfo("Save to storage \(passPhrase.longid)")
            storage.addPassPhrase(object: PassPhraseObject(passPhrase))
        } else {
            logger.logInfo("Save locally \(passPhrase.longid)")

            let locallPassPhrase = LocalPassPhrase(passPhrase: passPhrase, date: Date())
            localStorage.save(passPhrase: locallPassPhrase)

            let alreadySaved = storage.getPassPhrases()

            if alreadySaved.contains(where: { $0.longid == passPhrase.longid }) {
                storage.removePassPhrase(object: PassPhraseObject(passPhrase))
            }
        }
    }
```

So if it's not in storage - then it `ask` `localStorage` to save it. Local storage is a shared instance(live only during lifetime of application), it doesn't have any persisted storages inside or any else. It just have one in memory `private(set) var passPhrases: Set<LocalPassPhrase> = []`

Reply:
Cool, didn't know. Yes, indeed it returns `needPassphrase`. Fixed